File: P3/Napakalaki/src/napakalaki/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
/**
 *
 * @author 
 */
public class CardDealer {
    private static CardDealer instance=null;
    private ArrayList<Monster> unusedMonsters=new ArrayList();
    private ArrayList<Monster> usedMonsters=new ArrayList();
    private ArrayList<Treasure> unusedTreasures=new ArrayList();
    private ArrayList<Treasure> usedTreasures=new ArrayList();
    
    private CardDealer(){
    }

    public ArrayList<Monster> getUnusedMonsters() {
        return unusedMonsters;
    }

    public ArrayList<Monster> getUsedMonsters() {
        return usedMonsters;
    }

    public ArrayList<Treasure> getUnusedTreasures() {
        return unusedTreasures;
    }

    public ArrayList<Treasure> getUsedTreasures() {
        return usedTreasures;
    }
    
    private void initTreasureCardDeck(){
        unusedTreasures = new ArrayList();
        
        unusedTreasures.add(new Treasure("¡Sí mi amo!", 4, TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Botas de investigación", 3, TreasureKind.SHOES));
        unusedTreasures.add(new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("A prueba de babas", 2, TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Botas de lluvia ácida", 1, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Casco minero", 2, TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Ametralladora Thompson", 4, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Camiseta de la UGR", 1, TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Fez alópodo", 3, TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Hacha prehistórica", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Gaita", 4, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Insecticida", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Garabato Mistico", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("La rebeca metálica", 2, TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-comicón", 1, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necronomicón", 5, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Linterna a 2 manos", 3, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-gnomicón", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necrotelecom", 2, TreasureKind.HELMET));        
        unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necro-playboycon", 3, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND));       
        unusedTreasures.add(new Treasure("Tentácula de pega", 2, TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES));

    }
    
    private void initMonsterCardDeck(){
        unusedMonsters = new ArrayList();
        BadConsequence bc;
        Prize pz;
      
        ////////////////////////////////////////////////////////////////////////
      
        bc = new BadConsequence("Pierdes 5 niveles y 3 tesoros visibles", 5,3,0);
        pz = new Prize(4,2);
        unusedMonsters.add(new Monster("El rey de rosa",13,bc,pz ));
      
        ////////////////////////////////////////////////////////////////////////
      
        bc = new BadConsequence("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta"
          ,0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList(Arrays.asList(TreasureKind.ONEHAND)));
        
        pz=new Prize(4,1);
        unusedMonsters.add(new Monster("Ángeles de la noche ibicenca",14, bc, pz));
           
        
        ////////////////////////////////////////////////////////////////////////
        
        
        bc = new BadConsequence("Pierdes tu arma visible y otra oculta",
        0,new ArrayList(Arrays.asList(TreasureKind.ARMOR)),new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        pz = new Prize(2,1);
        unusedMonsters.add(new Monster("Byakhees de bonanza", 8, bc, pz));
        
        
        ////////////////////////////////////////////////////////////////////////
        
        bc= new BadConsequence("Embobados con el lindo primigenio te descartas de tu casco visible",
        0,new ArrayList(Arrays.asList(TreasureKind.HELMET)), new ArrayList());
        pz = new Prize(1,1);
        unusedMonsters.add(new Monster("Chibithulhu", 2, bc, pz));
        
        
        ////////////////////////////////////////////////////////////////////////
        
        bc= new BadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible",
        0,new ArrayList(Arrays.asList(TreasureKind.SHOES)), new ArrayList());
        pz=new Prize(1,1); 
        unusedMonsters.add(new Monster("El sopor de Dunwich", 2, bc,pz));

        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Pierde todos tus tesoros visibles ", 0, 5,0);
        pz=new Prize(3,1);
        unusedMonsters.add(new Monster("El gorron en el umbral",10,bc,pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Pierdes la armadura visible",0,new ArrayList(Arrays.asList(TreasureKind.ARMOR)),new ArrayList());
        pz=new Prize(2,1);
        unusedMonsters.add(new Monster("H.P. Munchcraft", 6, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible", 0,new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList());
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("Bichgooth", 2, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Toses los pulmones y pierdes 2 niveles", 2, 0, 0);
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("La que redacta las tinieblas", 2, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto",
        true);
        pz=new Prize(2,1);
        unusedMonsters.add(new Monster("Los Hondos",8, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Pierdes 2 niveles y 2 ocultos.", 0, 0, 2);
        pz=new Prize(2,1);
        unusedMonsters.add(new Monster("Semillas Cthulhu", 4, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Te intentas escaquear.Pierdes una mano visible.",0,new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),new ArrayList());
        pz=new Prize(2,1);
        unusedMonsters.add(new Monster("Dameargo", 1, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Da mucho asquito. Pierdes 3 niveles ",3,0,0 );
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("Pollipolipo volante", 3,bc,pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("No le hace gracia que pronuncien mal su nombre. Estas muerto", true);
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("YskhtihyssgGoth", 12, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("La familia te atrapa. Estas muerto" ,true);
        pz=new Prize(4,1);
        unusedMonsters.add(new Monster("La familia feliz", 1, bc, pz));
        
        ////////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("La quinta directiva primaria te obliga a perder 2 niveles un tesoro 2 manos visible", 2, 
                new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList());
        pz=new Prize(2,1);
        unusedMonsters.add(new Monster("Roboggoth", 8,bc,pz));
        
        ///////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Te asusta en la noche. Pierdes un casco visible.", 0,new ArrayList(Arrays.asList(TreasureKind.HELMET))
                ,new ArrayList());
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("El espía", 5,bc,pz));
        
        ///////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Menudo susto te llevas.Pierdes 2 niveles y 5 tesoros visibles.", 2,5,0);
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("El Lenguas", 20, bc, pz));
        
        /////////////////////////////////////////////////////////////////////
        
        bc=new BadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos.",3,
                new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList());
        pz=new Prize(1,1);
        unusedMonsters.add(new Monster("Bicéfalo", 20, bc, pz));
        
    }
    
    private void shuffleTreasures(){
        Collections.shuffle(unusedTreasures);
    }
    
    private void shuffleMonsters(){
        Collections.shuffle(unusedMonsters);
    }
    
    public static CardDealer getInstance()
    {
        if(instance==null)
            instance=new CardDealer();
        
        return instance;   
    }
    
    public Treasure nextTreasure(){   // MODIFICADO COPY
        //1- Comprobamos si el mazo de cartas sin usar esta vacio
        if (this.unusedTreasures.isEmpty()) 
        {            
            //Si esta vacio recorremos las cartas usadas y las convertimos a sin usar
            //la agregamos al mazo de cartas -> unusedTreasures
            for (Treasure t: this.usedTreasures) {
                this.unusedTreasures.add(t);
            }            
            //Barajamos el mazo con el metodo siguiente
            shuffleTreasures();
            
            //Limpiamos el mazo de descartes
            this.usedTreasures.clear(); //clear -> Removes all of the elements from this list        
        } 
        //2-
        //Sacamos la primera carta del mazo de cartas sin usar
        Treasure t = this.unusedTreasures.get(0);        
        //Metemos la carta en el mazo de cartas usadas
        this.usedTreasures.add(t);        
        //Eliminamos la carta del mazo de cartas sin usar
        this.unusedTreasures.remove(t);        
        //Mostramos la carta
        return t;
    }
    
    public Monster nextMonster(){   // MODIFICADO COPY
        //Comprobamos si el mazo de cartas sin usar esta vacio
        if (this.unusedMonsters.isEmpty())
        {            
            //Si esta vacio recorremos las cartas usadas y las convertimos a sin usar
            //la agregamos al mazo de cartas -> unusedMonsters
            for (Monster m: this.usedMonsters) {
                this.unusedMonsters.add(m);
            }            
            //Barajamos el mazo con el metodo siguiente
            shuffleMonsters();
            
            //Limpiamos el mazo de descartes
            this.usedMonsters.clear(); ////clear -> Removes all of the elements from this list         
        }        
        //Sacamos la primera carta del mazo de cartas sin usar
        Monster m = this.unusedMonsters.get(0);        
        //Metemos la carta en el mazo de cartas usadas
        this.usedMonsters.add(m);        
        //Eliminamos la carta del mazo de cartas sin usar
        this.unusedMonsters.remove(m);        
        
        return m;
    }
    
    public void giveTreasureBack(Treasure t){
        this.usedTreasures.add(t);
    }
    
    public void giveMonsterBack(Monster m){
        this.usedMonsters.add(m);
    }
    
    public void initCards()
    {
        this.initTreasureCardDeck();
        this.initMonsterCardDeck();
    }
    
    
}

File: P3/NapakalakiRuby/lib/card_dealer.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module NapakalakiRuby
  
  require "singleton" #modificado
  require_relative 'treasure'
  require_relative 'treasure_kind'
  require_relative 'monster'
  
  class CardDealer

    include Singleton
    attr_reader  :unusedMonsters, :usedMonsters, :usedTreasures, :unusedTreasures

    #@@instance=nil

    def initialize
      @unusedMonsters=Array.new
      @usedMonsters=Array.new
      @usedTreasures=Array.new
      @unusedTreasures=Array.new
    end
    

    def initTreasureCardDeck
     # @unusedTreasures = Array.new

      @unusedTreasures << Treasure.new("Si mi amo", 4, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Botas de investigacion", 3, TreasureKind::SHOES);
      @unusedTreasures << Treasure.new("Capucha de Cthulhu", 3, TreasureKind::HELMET);
      @unusedTreasures << Treasure.new("A prueba de babas", 2, TreasureKind::ARMOR);
      @unusedTreasures << Treasure.new("Botas de lluvia acida", 1, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Casco minero", 2, TreasureKind::HELMET);
      @unusedTreasures << Treasure.new("Ametralladora Thompson", 4, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Camiseta de la UGR", 1, TreasureKind::ARMOR);
      @unusedTreasures << Treasure.new("Clavo de rail ferroviario", 3, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Cuchillo de sushi arcano", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Fez alopodo", 3, TreasureKind::HELMET);
      @unusedTreasures << Treasure.new("Hacha prehistorica", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("El aparato del Pr. Tesla", 4, TreasureKind::ARMOR);
      @unusedTreasures << Treasure.new("Gaita", 4, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Insecticida", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Escopeta de 3 canones", 4, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Garabato Mistico", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("La rebeca metalica", 2, TreasureKind::ARMOR);
      @unusedTreasures << Treasure.new("Lanzallamas", 4, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Necro-comicon", 1, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Necronomicon", 5, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Linterna a 2 manos", 3, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Necro-gnomicon", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Necrotelecom", 2, TreasureKind::HELMET);        
      @unusedTreasures << Treasure.new("Mazo de los antiguos", 3, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Necro-playboycon", 3, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Porra preternatural", 2, TreasureKind::ONEHAND);
      @unusedTreasures << Treasure.new("Shogulador", 1, TreasureKind::BOTHHANDS);
      @unusedTreasures << Treasure.new("Varita de atizamiento", 3, TreasureKind::ONEHAND);       
      @unusedTreasures << Treasure.new("Tentacula de pega", 2, TreasureKind::HELMET);
      @unusedTreasures << Treasure.new("Zapato deja-amigos", 1, TreasureKind::SHOES); 
    
    end

    def initMonsterCardDeck
      @unusedMonsters = Array.new

      #1-Byakhees de bonanza
      prize = Prize.new(2,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Pierdes tu armadura visible y otra oculta',0,[TreasureKind::ARMOR],[TreasureKind::ARMOR])
      @unusedMonsters << Monster.new('3 Byakhees de bonanza',8, badconsequence, prize)

      #2-Chibithulhu
      prize = Prize.new(1,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Embobados con el lindo primigenio te descartas de tu casco visible', 0, [TreasureKind::HELMET], Array.new)
      @unusedMonsters << Monster.new('Chibithulhu',2, badconsequence, prize)

      #3-El sopor de Dunwich
      prize = Prize.new(1,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('El primordial bostezto contagioso. Pierdes el calzado visible',0,[TreasureKind::SHOES], Array.new)
      @unusedMonsters << Monster.new('El sopor de Dunwich', 2,badconsequence, prize)

      #4-Angeles de la noche ibicenca
      prize = Prize.new(4,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo.Descarta 1 mano visible y 1 mano oculta', 0, [TreasureKind::ONEHAND], [TreasureKind::ONEHAND])
      @unusedMonsters << Monster.new('Angeles de la noche ibicenca',14,badconsequence, prize)

      #5-El gorron en el umbral
      prize = Prize.new(3,1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes todos tus tesoros visibles', 0, 10, 0)
      @unusedMonsters << Monster.new('El gorron en el umbral', 10, badconsequence, prize)

      #6-H.P. Munchcraft
      prize = Prize.new(2,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Pierdes la armadura visible', 0, [TreasureKind::ARMOR], Array.new)
      @unusedMonsters << Monster.new('H.P. Munchcraft', 6, badconsequence, prize)

      #7-Bichgooth
      prize = Prize.new(1, 1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Sientes bichos bajo la ropa', 0, [TreasureKind::ARMOR], Array.new)
      @unusedMonsters << Monster.new('Bichgooth', 2, badconsequence, prize)

      #8-El rey de la rosa
      prize = Prize.new(4,2)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes 5 niveles y 3 tesoros visibles',5,3,0)
      @unusedMonsters << Monster.new('El rey de la rosa', 13, badconsequence, prize)

      #9-La que redacta en las tinieblas
      prize = Prize.new(1,1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Toses los pulmones y pierdes 2 niveles', 2, 0, 0)
      @unusedMonsters << Monster.new('La que redacta en las tinieblas', 2, badconsequence, prize)

      #10-Los hondos
      prize = Prize.new(2,1)
      badconsequence = BadConsequence.newDeath('Estos monstruos resultan bastante superficales y te aburren mortalmente. Estas muerto')
      @unusedMonsters << Monster.new('Los hondos',8,badconsequence, prize)

      #11-Semillas Cthulhu
      prize = Prize.new(2,1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes 2 niveles y 2 tesoros ocultos', 2, 0, 2)
      @unusedMonsters << Monster.new('Semillas Cthulhu', 4, badconsequence,prize)

      #12-Dameargo
      prize = Prize.new(2,1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Te intentas escaquear. Pierdes una mano visible', 0, [TreasureKind::ONEHAND], 0)
      @unusedMonsters << Monster.new('Dameargo',1, badconsequence, prize)

      #13-Pollipolipo volante
      prize = Prize.new(1,1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Da mucho asquito. Pierdes 3 niveles', 3, 0, 0)
      @unusedMonsters << Monster.new('Pollipolipo volante', 3, badconsequence,prize)

      #14-Yskhtihyssg-Goth
      prize = Prize.new(3,1)
      badconsequence = BadConsequence.newDeath('No le hace mucha gracia que pronuncien mal su nombre. Estas muerto')
      @unusedMonsters << Monster.new('Yskhtihyssg-Goth', 12, badconsequence,prize)

      #15-La familia feliz
      prize = Prize.new(4,1)
      badconsequence = BadConsequence.newDeath('La familia te atrapa. Estas muerto')
      @unusedMonsters << Monster.new('La familia feliz', 1, badconsequence,prize)

      #16-Roboggoth
      prize = Prize.new(2, 1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('La quinta directiva primaria te obliga a perder 2 niveles y un tesoro de 2 manos visible', 2, [TreasureKind::BOTHHANDS], 0)
      @unusedMonsters << Monster.new('Roboggoth', 8, badconsequence, prize)

      #17-El espia
      prize = Prize.new(1, 1)
      badconsequence = BadConsequence.newLevelSpecificTreasures('Te asusta en la noche. Pierdes un casco visible', 0, [TreasureKind::HELMET], 0)
      @unusedMonsters << Monster.new('El espia', 5, badconsequence,prize)

      #18-El lenguas
      prize = Prize.new(1, 1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles', 2, 5, 0)
      @unusedMonsters << Monster.new('El lenguas', 20, badconsequence, prize)

      #19-Bizefalo
      prize = Prize.new(1,1)
      badconsequence = BadConsequence.newLevelNumberOfTreasures('Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos', 3, 10, 0)
      @unusedMonsters << Monster.new('Bizefalo', 20, badconsequence, prize)

    end

    def shuffleTreasures
      @unusedTreasures = @unusedTreasures.shuffle
    end

    def shuffleMonsters
      @unusedMonsters = @unusedMonsters.shuffle
    end

    def nextTreasure
     #1- Comprobamos si el mazo de cartas sin usar esta vacio
        if @unusedTreasures.empty?

  #        Si esta vacio recorremos las cartas usadas y las convertimos a sin usar
  #        la agregamos al mazo de cartas -> unusedTreasures
          @usedTreasures.each do |t|                 

          @unusedTreasures<<t

          end

          #Barajamos el mazo con el metodo siguiente
          shuffleTreasures

          #Limpiamos el mazo de descartes
          @usedTreasures.clear

        end
        #2-Sacamos la primera carta del mazo de cartas sin usar
        t = @unusedTreasures.at(0)        
        #Metemos la carta en el mazo de cartas usadas
        @usedTreasures<<t        
        #Eliminamos la carta del mazo de cartas sin usar
        @unusedTreasures.delete(t);        

        return t
    end

    def nextMonster
      if @unusedMonsters.empty?            
  #        Si esta vacio recorremos las cartas usadas y las convertimos a sin usar
  #        la agregamos al mazo de cartas -> unusedMonsters
          @usedMonsters.each do |m|                 

          @unusedMonsters<<m

          end

          #Barajamos el mazo con el metodo siguiente
          shuffleMonsters

          #Limpiamos el mazo de descartes
          @usedTreasures.clear

        end

        #Sacamos la primera carta del mazo de cartas sin usar
        m = @unusedMonsters.at(0)

        #Metemos la carta en el mazo de cartas usadas
        @usedMonsters<<m

        #Eliminamos la carta del mazo de cartas sin usar
        @unusedMonsters.delete(m);

        return m 
    end

    def giveTreasureBack(treasure)
      @usedTreasures << treasure
    end

    def giveMonsterBack(monster)
      @usedMonsters << monster
    end

    def initCards
      initTreasureCardDeck
      initMonsterCardDeck
    end
  end
end
